### DISCUSSION TITLE
Proposal: add more array methods to Node for navigating children

### DISCUSSION THREAD

**Discussion by mattvague** [josdejong/mathjs#2589] [Ideas]
Right now we have `Node.forEach`, `Node.map`, etc which are very useful for navigating and modifying node children, however I find myself doing things like `node.args.some(...)`, `node.args.every(...)` fairly often and it'd be a bit nicer to be able to do `node.some(...)`, `node.every(...)` instead.

Relatedly, I wonder if we might actually consider rewriting the existing Array-like methods on Node using a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/Proxy). This would mean we're relying on Javascript's internal implementation rather than having to reimplement each ourselves. A good example might be https://www.30secondsofcode.org/articles/s/javascript-object-array-proxy

**Comment by josdejong** [comment]
Yes good idea! I have the same issues, `traverse` and `transform` are not always sufficient, and requires to write quite some plumbing to get things working.

I would love to have the following methods:

- `find`
- `some`
- `every`

I'm not sure yet if I would like to have them both recursive or shallow, or only recursive. I feel like only recursive would be best.

Can you explain a bit more about your idea with a proxy? Right now every different node class has one or a few properties that need to be iterated. For example `FunctionNode.prototype.map` has to iterate over both `this.fn` (a Node or string) and `this.args` (an array with Nodes), so that needs some custom code to handle. It is possible to write code to automatically iterate over all properties of any Node, but I do like to have control there, and explicitly iterate over the properties in a specified order.
